src/app/dao/tests/PartDAO.spec.ts

```diff
@@ -60,43 +84,363 @@ describe('PartDAO', () => {
+            await userDAO.update(creator.id, { observedPart: partId, last_changed });
+            await signOut();
+
+            // and given another user
+            await createConnectedUser(MALICIOUS_EMAIL, MALICIOUS_NAME);
+
+            // When the other user deletes the part
+            const result: Promise<void> = partDAO.delete(partId);
+
+            // Then it should fail
+            await expectPermissionToBeDenied(result);
+        });
+        it('should allow starting a part when chosen as opponent', async() => {
+            // Given a part, an accepted config, and a user who is the chosen opponent in the joiner
+            // Creator creates the joiner
+            const creator: MinimalUser = await createConnectedUser(CREATOR_EMAIL, CREATOR_NAME);
+            const part: Part = { ...PartMocks.INITIAL, playerZero: creator };
+            const partId: string = await partDAO.create(part);
+            await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+            await signOut();
+
+            // A candidate adds themself to the candidates list
+            const candidate: MinimalUser = await createConnectedUser(CANDIDATE_EMAIL, CANDIDATE_NAME);
+            await expectAsync(joinerDAO.addCandidate(partId, candidate)).toBeResolvedTo();
+            await signOut();
+
+            // The creator then selects candidates as the chosen opponent and proposes the config
+            await reconnectUser(CREATOR_EMAIL);
+            const update: Partial<Joiner> = {
+                chosenOpponent: candidate,
+                partStatus: PartStatus.CONFIG_PROPOSED.value,
+            };
+            await expectAsync(joinerDAO.update(partId, update)).toBeResolvedTo();
+            await signOut();
+
+            // And the candidate accepts the config
+            await reconnectUser(CANDIDATE_EMAIL);
+            await joinerDAO.update(partId, { partStatus: PartStatus.PART_STARTED.value });
+
+            // When chosen opponents updates the part document
+            const result: Promise<void> = partDAO.updateAndBumpIndex(partId,
+                                                                     Player.ONE,
+                                                                     part.lastUpdate.index,
+                                                                     {
+                                                                         playerZero: UserMocks.CREATOR_MINIMAL_USER,
+                                                                         playerOne: UserMocks.CANDIDATE_MINIMAL_USER,
+                                                                         turn: 0,
+                                                                         beginning: serverTimestamp(),
+                                                                         remainingMsForZero: 1000,
+                                                                         remainingMsForOne: 1000,
```

Comment:
where are those two from, almost look like candidate just imposed them to creator no ?

Reply:
tiens oui c'est clairement imposé par le candidat, j'ai rajouté une règle et deux tests pour éviter cela !

Reply:
Haha trop bien, comme quoi l'unit-test fait découvrir aussi les feature manquantes !